server/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -201,13 +240,138 @@ public ScriptService(Settings settings, Map<String, ScriptEngine> engines, Map<S
+            throw new IllegalArgumentException("Invalid contexts for script cache settings [" + String.join(", ", badContexts) + "]");
+        }
+
+        for (String contextName: contexts.keySet()) {
+            contextCache.put(contextName,
+                new ScriptCache(
+                    sizePerContext.getOrDefault(contextName, SCRIPT_CACHE_SIZE_SETTING.getDefault(settings)),
+                    expirePerContext.getOrDefault(contextName, SCRIPT_CACHE_EXPIRE_SETTING.getDefault(settings)),
                     compilationLimitsEnabled() ?
-                        SCRIPT_MAX_COMPILATIONS_RATE.get(settings):
+                        ratePerContext.getOrDefault(contextName, SCRIPT_MAX_COMPILATIONS_RATE.getDefault(settings)) :
                         new Tuple<>(0, TimeValue.ZERO)
-                    );
+                )
+            );
+        }
+
+        return null;
+    }
+
+    /**
+     * Return all setting values that do not correspond to the set of contexts.  Visible for testing.
+     */
+    static List<String> invalidContextSettings(Set<String> contextNames, Set<String> size, Set<String> expire, Set<String> rate) {
+        Set<String> badSize = new HashSet<>(size);
+        Set<String> badExpire = new HashSet<>(expire);
+        Set<String> badRate = new HashSet<>(rate);
+        for (String context: contextNames) {
+            badSize.remove(context);
+            badExpire.remove(context);
+            badRate.remove(context);
+        }
+
+        List<String> badSettings = new ArrayList<>();
+        for (String sizeContext: badSize) {
+            badSettings.add(SCRIPT_CACHE_SIZE_SETTING.getConcreteSettingForNamespace(sizeContext).getKey());
+        }
+        for (String expireContext: badExpire) {
+            badSettings.add(SCRIPT_CACHE_EXPIRE_SETTING.getConcreteSettingForNamespace(expireContext).getKey());
+        }
+        for (String rateContext: badRate) {
+            badSettings.add(SCRIPT_MAX_COMPILATIONS_RATE.getConcreteSettingForNamespace(rateContext).getKey());
+        }
+
+        Collections.sort(badSettings);
+        return badSettings;
+    }
+
+    /**
+     * Ensure that either the general cache settings are given or the per context settings are given.  If neither are set, the default is to
+     * using general cache settings.
+     * Visible for testing
+     * TODO(stu): change default to per context when context defaults exist
+     */
+    static boolean useGeneralCacheSettings(Settings settings) {
+        Set<String> generalKeys = new HashSet<>();
+        Set<String> contextKeys = new HashSet<>();
+        for (String key: settings.keySet()) {
+            if (SCRIPT_CACHE_SIZE_SETTING_DEPRECATED.match(key) ||
+                SCRIPT_CACHE_EXPIRE_SETTING_DEPRECATED.match(key) ||
+                SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.match(key)) {
+
+                generalKeys.add(key);
+
+            } else if (
+                SCRIPT_CACHE_SIZE_SETTING.getRawKey().match(key) ||
```

Comment:
This looks like the case for cross setting validation. Whether the general cache should be created should strictly be based on any general cache settings existing. Validating both sets of settings don't exist at the same time is a separate issue, and can be done by the settings validators themselves.

Reply:
> Validating both sets of settings don't exist at the same time is a separate issue, and can be done by the settings validators themselves.

Let's discuss how this can be done.

Reply:
We're no longer worrying about cross validation.  Now we allow any combo of settings to be set and flip between context and general caches using `script.max_compilations_rate=use-context`.